src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,373 @@
+        {
+            if (this.scopeMetrics.TryGetValue(metric.MeterName, out var metricList))
+            {
+                metricList.Add(metric);
+            }
+            else
+            {
+                if (MetricListPool.TryTake(out var newList))
+                {
+                    newList.Add(metric);
+                    this.scopeMetrics[metric.MeterName] = newList;
+                }
+                else
+                {
+                    newList = new List<Metric>();
+                    newList.Add(metric);
+                    this.scopeMetrics[metric.MeterName] = newList;
+                }
+            }
+        }
+
+        // Serialize
+        this.SerializeResourceMetrics(buffer, ref currentPosition, resource);
+
+        this.ReturnScopeMetrics();
+    }
+
+    internal void ReturnScopeMetrics()
+    {
+        foreach (var list in this.scopeMetrics)
+        {
+            list.Value.Clear();
+            MetricListPool.Add(list.Value);
+        }
+
+        this.scopeMetrics.Clear();
+    }
+
+    internal void SerializeResourceMetrics(byte[] buffer, ref int currentPosition, Resource resource)
+    {
+        this.resourceMetricStartIndex = currentPosition;
+
+        currentPosition += LengthAndTagSize;
+
+        // Serialize Resource
+        // TODO: Avoid serializing it multiple times.
+        SerializeResource(buffer, ref currentPosition, resource);
+
+        // TODO: Serialize schema_url field
+
+        // Serialize ScopeMetrics field
+        foreach (KeyValuePair<string, List<Metric>> metric in this.scopeMetrics)
+        {
+            this.previousScopeMetricsStartIndex = currentPosition;
+
+            currentPosition += LengthAndTagSize;
+
+            // Serialize this meter/scope
+            this.SerializeScopeMetrics(buffer, ref currentPosition, metric.Key, metric.Value);
+
+            // Reset the current position to start writing again from where we started the last scope.
+            currentPosition = this.previousScopeMetricsStartIndex;
+        }
+    }
+
+    internal void SerializeScopeMetrics(byte[] buffer, ref int currentPosition, string scopeName, List<Metric> metrics)
+    {
+        bool scopeInformationIncluded = false;
+
+        foreach (Metric metric in metrics)
+        {
+            if (!scopeInformationIncluded)
+            {
+                // Serialize scope information
+                SerializeInstrumentationScope(buffer, ref currentPosition, scopeName, metric.MeterTags);
+
+                scopeInformationIncluded = true;
+            }
+
+            this.previousMetricStartIndex = currentPosition;
+
+            currentPosition += LengthAndTagSize;
+
+            // Serialize metrics for the meter/scope
+            this.SerializeMetric(buffer, ref currentPosition, metric);
+
+            // Reset the current position to start writing again from where we started the last metric.
+            currentPosition = this.previousMetricStartIndex;
```

Comment:
This would override the previously serialized metric data, right?

Reply:
Ah we are only serializing one MetricPoint at a time! That essentially means that this method should always start writing at the same index. It would be more readable to have the variable `previousMetricStartIndex` to `metricStartIndex` and have that populated in the caller of this method instead of doing this everytime:

```csharp
this.previousScopeMetricsStartIndex = currentPosition;
...
currentPosition = this.previousScopeMetricsStartIndex;
```

Reply:
Please see updated, is this what you were expecting?

Reply:
I was trying to avoid repeated calculations of the same thing. We are doing this for each scope metrics instance:
```csharp
currentPosition = this.scopeMetricsStartIndex;
currentPosition += LengthAndTagSize;
```

Something like this:

```csharp
this.scopeMetricsLengthAndTagIndex = currentPosition;
this.scopeMetricsValueIndex = currentPosition + LengthAndTagSize;

foreach (KeyValuePair<string, List<Metric>> metric in this.scopeMetrics)
{
    if (metric.Value.Count > 0)
    {
        // Serialize this meter/scope
        this.SerializeScopeMetrics(buffer, ref this.scopeMetricsValueIndex, metric.Key, metric.Value);
    }
}
```

Reply:
Also, rename those variables to `TagAndLength` instead of `LengthAndTag`, if the we first encode Tag and then the length 😄